### PR TITLE
Overhaul display name semantics

### PR DESCRIPTION
**Pull request by sbrannen**
Previously there was considerable confusion with regard to the
differences between names and display names within JUnit 5.

This pull request addresses these issues by consolidating the two
competing concepts into a single "display name" concept. Test classes
and test methods may be assigned custom display names via the
`@DisplayName` annotation. Otherwise, a default display name will be
generated.

Specifically, the default display name for a test class is the fully
qualified name of the class.

The default display name for a test method is the name of the method
concatenated with a comma-separated list of parameter types in
parentheses, and the names of parameter types are retrieved using
`Class#getSimpleName()`. For example, the default display name for the
following test method is `testUser(TestInfo, User)`.

```java
  @Test
  void testUser(TestInfo testInfo, @Mock User user) { ... }
```

Issue: #153

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

### CODE REVIEW

**Comment by jlink** [comment]
Using FQCN as default display name for test classes makes it difficult for test writers to just make some test classes more readable via "@DisplayName" since the more readable version should also contain the package(s). Otherwise the mixture of long and short names will look either strange in the IDE or the IDE will have to display it differently depending on if the package(s) are present or not.

**Comment by marcphilipp** [comment]
@jlink I agree. However, since packages are not present in the test tree at the moment it might be confusing to use simple names for classes. I think we could use the simple names only if we added packages (as @sbrannen [suggested in #153](https://github.com/junit-team/junit5/issues/153#issuecomment-220847441)).

But I don't see us doing that for M1, especially because it's unclear how. There are at least two ways. For example for a class org.foo.SomeTest and org.foo.bar.AnotherTest the test tree could look like

1. Flat
  - org.foo
    - SomeTest
  - org.foo.bar
    - AnotherTest
2. Deep
  - org
    - foo
      - bar
        - AnotherTest
      - SomeTest

I'm undecided as to which one I would prefer.

**Comment by jlink** [comment]
Having the package now and remove it later in M2 will force the IDEs to change their display policy. If an IDE wants to display the package name it can always use the test source.

**Comment by sbrannen** [comment]
> If an IDE wants to display the package name it can always use the test source.

Yes, I think that's the crucial thing to keep in mind: an IDE or build tool can always use the `TestSource`.

That is in fact what I had to do in our XML report generator for this PR since `getName()` is no longer available. Consequently, we are now using the `JavaClassSource` to retrieve the FQCN for use in XML reports.

In general, IDEs and build tools should only use the _display name_ for display purposes, not for building a tree in the UI. From that perspective, it should not matter what is _in_ the display name.

However, I do see the point that using a FQCN as the default display name leads to rather long display names in the UI.

In fact, my original proposal in #153 stripped the package names from class names. So I am not opposed to doing so. Plus, as you mentioned @jlink, it is probably better to do it in M1 than changing it in M2. The only downside is that navigation back to a test class from the UI tree in IDEs likely will not work at all. But perhaps it's better to bite the bullet now.

@junit-team/junit-lambda, thoughts?

**Comment by sbrannen** [comment]
@marcphilipp, I think we should defer the decision regarding what the test tree will physically look like (i.e., in terms of flattening packages or not) until M2.

**Comment by marcphilipp** [comment]
> However, I do see the point that using a FQCN as the default display name leads to rather long display names in the UI.
> 
> In fact, my original proposal in #153 stripped the package names from class names. So I am not opposed to doing so. Plus, as you mentioned @jlink, it is probably better to do it in M1 than changing it in M2. The only downsides are that navigation back to a test class from the UI tree in IDEs likely will not work well, and the trees in UIs will be flat without the package name in the display name (at least in Eclipse). But perhaps it's better to bite the bullet now.

Ok, you've convinced me. I think we should use the simple name and document how to retrieve the FQCN in case IDEs want to display it in M1.

> I think we should defer the decision regarding what the test tree will physically look like (i.e., in terms of flattening packages or not) until M2.

Point taken. I was just trying to illustrate what I was talking about. 😉

**Comment by sbrannen** [comment]
Thanks for the feedback, guys!

I'll go ahead and switch to simple names for classes, except that I'll include the enclosing top-level class(es) for nested static classes (in line with my original proposal).

**Comment by marcphilipp** [comment]
👍

File: junit-tests/src/test/java/org/junit/gen5/console/tasks/TempDirectory.java [hunk 1]
```diff
@@ -71,7 +71,15 @@ private Namespace localNamespace(ExtensionContext context) {
 
 	private Path createTempDirectory(ExtensionContext context) {
 		try {
-			return Files.createTempDirectory(context.getName());
+			String tempDirName;
+			if (context.getTestMethod().isPresent()) {
+				tempDirName = context.getTestMethod().get().getName();
+			}
+			else {
+				tempDirName = context.getTestClass().get().getName();
```
Comment (marcphilipp) [inline junit-tests/src/test/java/org/junit/gen5/console/tasks/TempDirectory.java]: This call to `get()` strikes me as not being save in all cases. Can we use an additional fallback if it the test class is not present, e.g. `getDisplayName()`?

File: junit-console/src/main/java/org/junit/gen5/console/tasks/XmlReportWriter.java [hunk 2]
```diff
@@ -125,6 +126,17 @@ private void writeTestcase(TestIdentifier test, NumberFormat numberFormat, XMLSt
...
 	}
 
+	private String getParentClassName(TestIdentifier test) {
+		// @formatter:off
+		return reportData.getTestPlan().getParent(test)
+				.flatMap(TestIdentifier::getSource)
+				.filter(JavaClassSource.class::isInstance)
+				.map(JavaClassSource.class::cast)
+				.map(source -> source.getJavaClass().getName())
+				.orElse("<unrooted>");
```
Comment (marcphilipp) [inline junit-console/src/main/java/org/junit/gen5/console/tasks/XmlReportWriter.java]: Can we use the display name of the parent as a fallback? I think `<unrooted>` is really misleading. We can do it in a separate issue if you'd prefer.

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/JUnit5TestDescriptor.java [hunk 3]
```diff
@@ -55,12 +56,12 @@ protected JUnit5TestDescriptor(UniqueId uniqueId) {
...
 	}
 
-	protected String determineDisplayName(AnnotatedElement element, String defaultName) {
+	protected String determineDisplayName(AnnotatedElement element, Supplier<String> defaultNameGenerator) {
 		// @formatter:off
 		return findAnnotation(element, DisplayName.class)
 				.map(DisplayName::value)
 				.filter(StringUtils::isNotBlank)
-				.orElse(defaultName);
+				.orElse(defaultNameGenerator.get());
```
Comment (marcphilipp) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/JUnit5TestDescriptor.java]: I think this should be `orElseGet(defaultNameGenerator)`, right?

File: junit5-api/src/main/java/org/junit/gen5/api/DynamicTest.java [hunk 4]
```diff
@@ -40,31 +43,33 @@
 @API(Experimental)
 public class DynamicTest {
 
-	private final String name;
+	private final String displayName;
 	private final Executable executable;
 
 	public DynamicTest(String name, Executable executable) {
```
Comment (marcphilipp) [inline junit5-api/src/main/java/org/junit/gen5/api/DynamicTest.java]: Nitpick: Please also change the parameter name.

File: junit5-api/src/main/java/org/junit/gen5/api/DynamicTest.java [hunk 5]
```diff
@@ -40,31 +43,33 @@
...
-		this.name = name;
+		this.displayName = name;
 		this.executable = executable;
 	}
 
-	public String getName() {
-		return name;
+	public String getDisplayName() {
+		return this.displayName;
```
Comment (marcphilipp) [inline junit5-api/src/main/java/org/junit/gen5/api/DynamicTest.java]: Good catch BTW! 👍

**Comment by marcphilipp** [comment]
I commented a few minor things inline. Besides those, this PR looks really good! 👍

**Comment by sbrannen** [comment]
Update: I just pushed the changes for shorter default display names for test classes.

File: junit5-api/src/main/java/org/junit/gen5/api/DynamicTest.java (hunk 4 above)
Comment (sbrannen) [inline junit5-api/src/main/java/org/junit/gen5/api/DynamicTest.java]: done

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/JUnit5TestDescriptor.java (hunk 3 above)
Comment (sbrannen) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/JUnit5TestDescriptor.java]: Ummmm.... yeah.... except that the argument is now `Function<AnnotatedElement, String>`. 😉

File: junit-console/src/main/java/org/junit/gen5/console/tasks/XmlReportWriter.java (hunk 2 above)
Comment (sbrannen) [inline junit-console/src/main/java/org/junit/gen5/console/tasks/XmlReportWriter.java]: Well, this was the previous code:

```java
writer.writeAttribute("classname", parent.map(TestIdentifier::getName).orElse("<unrooted>"));
```

So I didn't want (and don't want) to change the existing behavior in this PR.

I assumed it was done that way for a reason.

In any case, feel free to raise a new issue if you think this should be changed.

File: junit-tests/src/test/java/org/junit/gen5/console/tasks/TempDirectory.java (hunk 1 above)
Comment (sbrannen) [inline junit-tests/src/test/java/org/junit/gen5/console/tasks/TempDirectory.java]: It's _only_ test code, but sure... I'll make it more robust. 😉

**Pull request merged**